pkg/ccl/sqlproxyccl/acl/watcher.go

```diff
@@ -0,0 +1,244 @@
+
+	// These control whether or not a connection is allowd based on it's ConnectionTags.
+	controllers []AccessController
+}
+
+// Listener contains the channel notified when a connection is denied
+// and contains state needed to remove the listener from the watcher.
+type listener struct {
+	// Lock Ordering: Only lock one listener at a time. Lock the listener before
+	// locking the watcher.
+	mu struct {
+		syncutil.Mutex
+
+		// The denied callback is notified iff the connection was blocked.
+		// After the callback is notified once, denied is set
+		// to nil to prevent duplicate calls.
+		denied func(error)
+	}
+
+	// Unique id. The id is required to disambiguate connections with identical
+	// connection tags.
+	id int64
+
+	// Used to identify if the connection matches the access control list.
+	connection ConnectionTags
+}
+
+// Option allows configuration of an access control list service.
+type Option func(*aclOptions)
+
+type aclOptions struct {
+	pollingInterval time.Duration
+	timeSource      timeutil.TimeSource
+	errorCount      *metric.Gauge
+	allowlistFile   string
+	denylistFile    string
+}
+
+// WithPollingInterval specifies interval between polling for config file
+// changes.
+func WithPollingInterval(d time.Duration) Option {
+	return func(op *aclOptions) {
+		op.pollingInterval = d
+	}
+}
+
+// WithTimeSource overrides the time source used to check expiration times.
+func WithTimeSource(t timeutil.TimeSource) Option {
+	return func(op *aclOptions) {
+		op.timeSource = t
+	}
+}
+
+func WithErrorCount(errorCount *metric.Gauge) Option {
+	return func(op *aclOptions) {
+		op.errorCount = errorCount
+	}
+}
+
+func WithAllowListFile(allowlistFile string) Option {
+	return func(op *aclOptions) {
+		op.allowlistFile = allowlistFile
+	}
+}
+
+func WithDenyListFile(denylistFile string) Option {
+	return func(op *aclOptions) {
+		op.denylistFile = denylistFile
+	}
+}
+
+func NewWatcher(ctx context.Context, opts ...Option) (*Watcher, error) {
+	options := &aclOptions{
+		pollingInterval: defaultPollingInterval,
+		timeSource:      timeutil.DefaultTimeSource{},
+	}
+	for _, opt := range opts {
+		opt(options)
+	}
+	w := &Watcher{
+		listeners:   btree.New(8),
+		options:     options,
+		controllers: make([]AccessController, 0),
+	}
+
+	if options.allowlistFile != "" {
+		c, next, err := newAccessControllerFromFile[*Allowlist](ctx, options.allowlistFile, w.options.pollingInterval, w.options.errorCount)
+		if err != nil {
+			return nil, err
+		}
+		w.addAccessController(c, next)
+	}
+	if options.denylistFile != "" {
+		c, next, err := newAccessControllerFromFile[*Denylist](ctx, options.denylistFile, w.options.pollingInterval, w.options.errorCount)
+		if err != nil {
+			return nil, err
+		}
+		w.addAccessController(c, next)
+	}
+	return w, nil
+}
+
+func (w *Watcher) addAccessController(controller AccessController, next chan AccessController) {
+	index := len(w.controllers)
+	w.controllers = append(w.controllers, controller)
```

Comment:
There is still a small race condition here. It's unlikely to be hit since the channel won't emit a new access controller for five seconds, but we should handle it. addAccessController does not lock controllers and updateAccessController copies the list.

Normally it is safe to assume the lock in the function constructing an object, but that assumption is broken if any of the methods called in the constructor spawns goroutines.

Reply:
It took me a second to see it, but you're totally right. Fixed!